File: Frameworks/mealpy/lpm/problems/Ackley.py
from mealpy import Problem
import numpy as np


class Ackley(Problem):
    def __init__(self, lb, ub, minmax, name="Ackley", **kwargs):
        super().__init__(lb, ub, minmax, **kwargs)
        self.name = name

    def fit_func(self, solution):
        a = 20.0
        b = 0.2
        c = 2 * np.pi
        sum1 = 0
        sum2 = 0
        for i in range(len(solution)):
            sum1 += solution[i] ** 2
            sum2 += np.cos(c * solution[i])
        return -a * np.exp(-b * np.sqrt(1.0 / len(solution) * sum1)) - np.exp(1.0 / len(solution) * sum2) + a + np.e

File: Frameworks/jMetal/jmetal-problem/src/main/java/org/uma/jmetal/problem/singleobjective/Ackley.java
package org.uma.jmetal.problem.singleobjective;

import org.uma.jmetal.problem.doubleproblem.impl.AbstractDoubleProblem;
import org.uma.jmetal.solution.doublesolution.DoubleSolution;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

import static java.lang.Math.*;

public class Ackley extends LoggingDoubleProblem {

    public Ackley() {
        this(10);
    }

    public Ackley(Integer numberOfVariables) {
        numberOfObjectives(1);
        name("Ackley");

        List<Double> lowerLimit = new ArrayList<>();
        List<Double> upperLimit = new ArrayList<>(numberOfVariables);

        IntStream.range(0, numberOfVariables).forEach(i -> {
            lowerLimit.add(-32.0);
            upperLimit.add(32.0);
        });

        variableBounds(lowerLimit, upperLimit);
    }

    @Override
    public DoubleSolution computeFitness(DoubleSolution solution) {
        double a = 20.0, b = 0.2, c = 2 * PI;
        double sum1 = 0, sum2 = 0;

        for (double v : solution.variables()) {
            sum1 += pow(v, 2);
            sum2 += cos(c * v);
        }
        solution.objectives()[0] = -a * exp(-b * sqrt(1.0 / numberOfVariables() * sum1)) - exp(1.0 / numberOfVariables() * sum2) + a + E;

        return solution;
    }
}
